src/service-worker/core.js

```diff
@@ -0,0 +1,309 @@
+ *
+ * @param {string} url
+ * @param {string} version
+ * @return {string}
+ */
+function versionedUrl(url, version) {
+  // Ensure we do not replace the prefix.
+  return url.replace(ampVersion(url), version);
+}
+
+/**
+ * Determines if a url is a request to a CDN JS file.
+ * @param {string} url
+ * @return {boolean}
+ */
+function isCdnJsFile(url) {
+  return url.endsWith('.js') && (
+    url.startsWith('https://cdn.ampproject.org/rtv') ||
+    url.startsWith('https://cdn.ampproject.org/v0')
+  );
+}
+
+
+/**
+ * A mapping from a Client's (unique per tab _and_ refresh) ID to the AMP
+ * release version we are serving it.
+ *
+ * @const
+ * @type {!Object<string, string>}
+ */
+const clientsMap = Object.create(null);
+
+/**
+ * Our cache of CDN JS files.
+ *
+ * @type {Cache}
+ */
+let cache;
+
+/**
+ * A (wrapped) IndexedDB Database, where we record which versions of a file
+ * have been cached. Used to determine if we have another version of the same
+ * file cached when we miss.
+ *
+ * @type {../../third_party/indexed-db-as-promised/classes/database.Database}
+ */
+let db;
+
+/**
+ * A promise to open up our CDN JS cache, which will be resolved before any
+ * requests are intercepted by the SW.
+ *
+ * @type {!Promise}
+ */
+const cachePromise = caches.open('cdn-js').then(result => {
+  cache = result;
+});
+
+/**
+ * A promise to open up our Indexed DB database, which will be resolved before
+ * any requests are intercepted by the SW.
+ *
+ * @type {!Promise}
+ */
+const dbPromise = indexedDBP.open('cdn-js', 1, {
+  upgrade(db, event) {
+    const oldVersion = event.oldVersion;
+    // Do we need to create our database?
+    if (oldVersion == 0) {
+      const files = db.createObjectStore('js-files', {autoIncrement: true});
+      files.createIndex('files', 'file');
+      files.createIndex('fileVersions', ['file', 'version']);
+    }
+  },
+}).then(result => {
+  db = result;
+});
+
+/**
+ * Fetches the request, and stores it in the cache. Since we only store one
+ * version of each file, we'll prune all older versions after we cache this.
+ *
+ * @param {!Request} request
+ */
+function fetchAndCache(request) {
+  const url = request.url;
+  const requestFile = basename(url);
+  const requestVersion = ampVersion(url);
+
+  // TODO(jridgewell): we should also fetch this requestVersion for all files
+  // we know about.
+
+  return fetch(request).then(response => {
+    // Did we receive a valid response (200 <= status < 300)?
+    if (response && response.ok) {
+      // You must clone to prevent double reading the body.
+      cache.put(request, response.clone());
+
+      // Store the file version in IndexedDB
+      // This intentionally does not block the request resolution to speed
+      // things up. This is likely fine since you don't have multiple
+      // `<script>`s with the same `src` on a page.
+      db.transaction('js-files', 'readwrite').run(transaction => {
```

Comment:
As discussed, you may want to look into whether you can avoid calling IndexedDB at all here, and if you need to store additional state information, use a "fake" URL query parameter in the cache keys. That *may* lead to some performance wins vs. having to open and interact with IndexedDB, but it would also mean not needing to pull in the `indexeddb-as-promised` dependency.